src/v/cluster/types.cc

```diff
@@ -293,24 +318,188 @@ std::ostream& operator<<(
+};
+
+/// Serialize tristate using the same format as adl
+///
+/// Template parameter T is a value type of the tristate.
+template<class T>
+inline void write(iobuf& out, const tristate<T>& t) {
+    if (t.is_disabled()) {
+        serde::write(out, static_cast<int8_t>(-1));
+    } else if (!t.has_value()) {
+        serde::write(out, static_cast<int8_t>(0));
+    } else {
+        serde::write(out, static_cast<int8_t>(1));
+        serde::write(out, t.value());
+    }
+}
+
+/// The wrapper is used to serialize topic_configuraiton using serde
+/// library.
+/// The topic_configuration doesn't have default c-tor so it can't be
+/// directly used with serde. Another benefit of using this wrapper is
+/// that serialization details (versions) are not exposed in the header.
+struct topic_configuration_wrapper {
+    using value_t = topic_configuration_wrapper;
+    static constexpr serde::version_t redpanda_serde_version = 0;
+    static constexpr serde::version_t redpanda_serde_compat_version = 0;
+    std::optional<cluster::topic_configuration> cfg;
+
+    void serde_read(iobuf_parser& in, const serde::header& hdr) {
+        auto ns = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+        auto tp = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+
+        auto partition_count = serde::read_nested<int32_t>(
+          in, hdr._bytes_left_limit);
+        auto replication_factor = serde::read_nested<int16_t>(
+          in, hdr._bytes_left_limit);
+
+        cfg = cluster::topic_configuration(
+          model::ns(ns), model::topic(tp), partition_count, replication_factor);
+
+        cfg->properties.compression
+          = serde::read_nested<std::optional<model::compression>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.cleanup_policy_bitflags
+          = serde::read_nested<std::optional<model::cleanup_policy_bitflags>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.compaction_strategy
+          = serde::read_nested<std::optional<model::compaction_strategy>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.timestamp_type
+          = serde::read_nested<std::optional<model::timestamp_type>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.segment_size
+          = serde::read_nested<std::optional<size_t>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.retention_bytes = serde::read_nested<tristate<size_t>>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.retention_duration
+          = serde::read_nested<tristate<std::chrono::milliseconds>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.recovery = serde::read_nested<std::optional<bool>>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.shadow_indexing
+          = serde::read_nested<std::optional<cluster::shadow_indexing_mode>>(
+            in, hdr._bytes_left_limit);
+    }
+
+    void serde_write(iobuf& out) {
```

Comment:
I would assume that this method is not necessary since it just lists the members in the order they occur in the struct itself. This should be handled by serde internally without writing additional code. If any types are not serde-compatible yet, this should be resolved by adding serialize/deserialize functions for those types.

The only reason to not use "auto-mode" of serde for struct serialization would be that this way you can reorder the members in the struct definition without changing the compat version. I would rather not reorder struct members and not write all those extra lines of code. But this is a design decision.